pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -631,13 +630,22 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 		m.allocatedDevices[resource] = sets.New[string]()
 	}
 
+	defer func() {
+		if err == nil {
+			if m.allocatedDevices[resource] == nil {
+				m.allocatedDevices[resource] = sets.New[string]()
+			}
+			m.allocatedDevices[resource] = m.allocatedDevices[resource].Union(allocated)
+		}
+	}()
```

Comment:
@payall4u Can you explain how and why this PR fixes concurrent access to the `m.allocatedDevices` from `allocateContainerResources` and  `UpdateAllocatedDevices` ?

Reply:
Sure. When call `callGetPreferredAllocationIfAvailable`, current goroutine when release the lock and call device plugin. 
If some call ListResource of resource server, devicemanager will recalculate the allocatedDevices map. It will remove some map entry if the resource has be released already.
https://github.com/kubernetes/kubernetes/blob/v1.28.0-beta.0/pkg/kubelet/apis/podresources/server_v1.go#L58

When return from `callGetPreferredAllocationIfAvailable`, the map entry is nil. So we shoud reinit the device set.

Reply:
I generally understood the idea, but it's still quite hard to grasp it. I agree with @ffromani, this code is asking for refactoring.